newrelic/hooks/mlmodel_langchain.py

```diff
@@ -255,6 +259,476 @@ def wrap_similarity_search(wrapped, instance, args, kwargs):
+    metadata = run_manager_info.get("metadata", {})
+    tags = run_manager_info.get("tags", "")
+
+    input_message_list = [_input]
+    output_message_list = []
+    response_len = 1
+    try:
+        output_message_list = [response[0]] if response else []
+    except:
+        try:
+            output_message_list = [str(response)]
+        except:
+            _logger.warning("Unable to capture response. No response message event will be captured.")
+
+    # Make sure the builtin attributes take precedence over metadata attributes.
+    full_chat_completion_summary_dict = {"metadata.%s" % key: value for key, value in metadata.items()}
+    full_chat_completion_summary_dict.update(
+        {
+            "id": completion_id,
+            "appName": app_name,
+            "conversation_id": conversation_id,
+            "span_id": span_id,
+            "trace_id": trace_id,
+            "transaction_id": transaction.guid,
+            "vendor": "langchain",
+            "ingest_source": "Python",
+            "virtual_llm": True,
+            "request_id": run_id,
+            "duration": ft.duration,
+            "response.number_of_messages": len(input_message_list) + len(output_message_list),
+            "tags": tags or "",
+        }
+    )
+
+    transaction.record_custom_event("LlmChatCompletionSummary", full_chat_completion_summary_dict)
+
+    create_chat_completion_message_event(
+        transaction,
+        settings.app_name,
+        input_message_list,
+        completion_id,
+        span_id,
+        trace_id,
+        run_id,
+        conversation_id,
+        output_message_list,
+        message_ids,
+    )
+
+    return return_val
+
+
+def wrap_chain_sync_run(wrapped, instance, args, kwargs):
+    transaction = current_transaction()
+    if not transaction:
+        return wrapped(*args, **kwargs)
+
+    # Framework metric also used for entity tagging in the UI
+    transaction.add_ml_model_info("Langchain", LANGCHAIN_VERSION)
+
+    run_args = bind_args(wrapped, args, kwargs)
+    # invoke has an argument named "config" that contains metadata and tags whereas
+    # run has arguments named "metadata" and "tags".
+    if "config" in run_args:
+        message_ids = ((run_args.get("config") or {}).get("metadata") or {}).pop("message_ids", [])
+    else:
+        message_ids = (run_args.get("metadata") or {}).pop("message_ids", [])
+    # invoke has an argument named "input" whereas the input when calling run can
+    # end up in args or kwargs.
+    _input = run_args.get("input", "") or run_args["kwargs"].get("input") or run_args["args"][0]
+
+    span_id = None
+    trace_id = None
+
+    # Get conversation ID off of the transaction
+    custom_attrs_dict = transaction._custom_params
+    conversation_id = custom_attrs_dict.get("llm.conversation_id", "")
+
+    settings = transaction.settings if transaction.settings is not None else global_settings()
+    app_name = settings.app_name
```

Comment:
I believe you only need to do that safeguard if you are only looking at the transaction.settings. global_settings can never be None. So since we have this transaction.settings or global_settings() logic we are guaranteed that they are not None. Same with the app_name in the settings. If the settings are not None we are guaranteed to have the app_name.